src/storage/src/source/mysql/replication.rs

```diff
@@ -0,0 +1,405 @@
+
+    let errors = definite_errors.concat(&transient_errors.map(ReplicationError::from));
+
+    (
+        replication_updates,
+        upper_stream,
+        errors,
+        button.press_on_drop(),
+    )
+}
+
+/// Produces the logical replication stream while taking care of regularly sending standby
+/// keepalive messages with the provided `uppers` stream.
+///
+/// The returned stream will contain all transactions that whose commit LSN is beyond `resume_lsn`.
+async fn raw_stream<'a>(
+    config: &RawSourceCreationConfig,
+    mut conn: Conn,
+    // TODO: this is only needed while we use a single source-id for the timestamp; remove once we move to a partitioned timestamp
+    gtid_source_id: Uuid,
+    resume_transaction_id: TransactionId,
+) -> Result<Result<BinlogStream, DefiniteError>, TransientError> {
+    // Verify the MySQL system settings are correct for consistent row-based replication using GTIDs
+    // TODO: Should these return DefiniteError instead of TransientError?
+    ensure_gtid_consistency(&mut conn).await?;
+    ensure_full_row_binlog_format(&mut conn).await?;
+    ensure_replication_commit_order(&mut conn).await?;
+
+    // To start the stream we need to provide a GTID set of the transactions that we've 'seen'
+    // and the server will send us all transactions that have been committed after that point.
+    // NOTE: The 'Gno' intervals in this transaction-set use an open set [start, end)
+    // interval, which is different than the closed-set [start, end] form returned by the
+    // @gtid_executed system variable. So the intervals we construct in this GTID set
+    // end with the value of the transaction-id that we want to start replication at,
+    // which happens to be our `resume_transaction_id`
+    // https://dev.mysql.com/doc/refman/8.0/en/replication-options-gtids.html#sysvar_gtid_executed
+    // https://dev.mysql.com/doc/dev/mysql-server/latest/classGtid__set.html#ab46da5ceeae0198b90f209b0a8be2a24
+    let seen_gtids = match resume_transaction_id.into() {
+        0 | 1 => {
+            // If we're starting from the beginning of the binlog or the first transaction id
+            // we haven't seen 'anything'
+            Vec::new()
+        }
+        ref a => {
+            // NOTE: Since we enforce replica_preserve_commit_order=ON we can start the interval at 1
+            // since we know that all transactions with a lower transaction id were monotonic
+            vec![Sid::new(*gtid_source_id.as_bytes()).with_interval(GnoInterval::new(1, *a))]
+        }
+    };
+
+    // Request that the stream provide us with a heartbeat message when no other messages have been sent
+    let heartbeat = Duration::from_secs(3);
+    conn.query_drop(format!(
+        "SET @master_heartbeat_period = {};",
+        heartbeat.as_nanos()
+    ))
+    .await?;
+
+    // Try to generate a deterministic server-id for identifying us as a replica on the upstream mysql server.
```

Comment:
This is a bit pedantic but "Try to" implies there is possibility of failure but I don't see any. We always successfully construct a deterministic server-id, right?

Reply:
Technically the construction could fail if the `GlobalId` + `REPLICATION_SERVER_ID_OFFSET` is greater than `u32::MAX`, which is why I worded it this way (since `GlobalId` is a `u64`). This seems unlikely to occur in practice since our ids appear to be fairly small, right?